src/main/scala/models/Customers.scala

```diff
@@ -0,0 +1,9 @@
+package models
```

Comment:
first of all - let's talk about foundations first, could you elaborate a little bit on
1. onion architecture
   - what is it all about
   - what package structure it suggests
   - what are the responsibilities of code contained in each package
   - what are the dependencies
      - which packages relies on which
      - which packages cannot be imported in which packages
1. hexagonal architecture
   - what is the purpose of it
   - what business problems it fixes
   - what is easier in that architecture
   - what is harder in that architecture
   - how it affects product maintainability / extensibility
   - how it affects testing
   - what are main difficulties
   - how it connects with onion architecture
      - how it affects package structure
      - what concepts it adds to onion architecture
1. tagless final
   - what is the main concept that it resolves
   - what are the drawbacks
   - how it affects onion & hexagon architecture

Reply:
I've never head of these concepts before so I will do my best to explain what I understand from my researches :

- Onion architecture aims to decouple the responsibilities of the application by making the core business logic independent from the rest of the application. It's suggesting the application to be separated into domain / application / infrastructure / UI.
Domain is responsible for containing the core business logic of the application, which include de model and the services.
Application is responsible to handle the business logic / implementation details
Infrastructure is responsible for interactions with the outer world, so db or api for example
UI, graphic interface
Domain should rely on none of the packages, application can rely on domain, infrastructure can rely on domain and application and UI can depend on everything

Hexagonal architecture aims to make the application more flexible and easily testable. The business problem solved is that implementation details can change but the core business logic can remain untouched.
Pros : Testing is easier because you can easily create mock for example, changing implementation details like the underlying db is also much easier, and there is a clear separation of concern so the code base might be easier to understand
Cons : Might be an overkill for small project because of the numerous abstractions. Implementation can also be tricky from what I've read (not too sure why)

Tagless final : 
It is solving the expression problem, which is adding new data variants or new operations without having to modify the existing code.
Cons : Can lead to boiler plate code, and debugging might be more difficult
It really fits into the onion & hexagon architecture because in both cases there is this notion of decoupling the reponsabilities, which tagless final is a perfect fit for

Reply:
- Onion architecture aims to decouple the responsibilities of the application by making the core business logic independent from the rest of the application - OK
- It's suggesting the application to be separated into domain / application / infrastructure / UI.
   - partly OK; instead of `application` -> `gateway`
   - gateway
      - translate outer world (like requests) into domain language (like commands (ex. createUserCommand(fields of user to be created)) or queries(getUsers(ids and limit)) )
      - translate domain language into outer world; domain service method is returning Customer; but for API u need json like this { customer_name, customer_surname, customer_id } instead of { name, surname, id }
      - either with annotations ( `@encodedName` ) or u can have completely separated concept like CustomerApiOutput (gateway package) vs Customer (domain package)

Reply:
- Domain is responsible for containing the core business logic of the application
   - Domain is responsible for ALL business logic
- which include de model and the services
   - model ~ domain concepts like User, Product, Customer
- Application is responsible to handle the business logic / implementation details
   - `Application` ~ `gateway`: it is only translating, not business logic, no implementation details
   - if service returns `Option.none` from `getCustomer`, u r translating it into 404 http status code
   - if service "throws" error like `customerNotFoundError`, u r translating it into 404 http status code

Reply:
- Infrastructure is responsible for interactions with the outer world, so db or api for example
UI
   - `infrastructure` is responsible for interactions with the outer world - that is true
   - `infrastructure` and `gateway` - what is the difference?
      - `infrastructure` contains implementations of repositories or in general adapters (implementations of ports)
      - `infrastructure` has logic to create services, repositories implementations etc
         - creating service is not always a matter of calling `new CustomerService(...)`, some service would need access to resources, for example some file
         - constructors are not enough, because u cannot handle errors => that why in ZIO u have all constructors transformed into Layers: ZLayer is like: ZLayer[Env, Error, Service]
             - `ZLayer[Env, Error, Service]` - ur creating `Service(service1, service2, repository1)` it means that `service1`, `service2`, `repository1` needs to be in `Env`
             - `ZLayer { service1 <- Service1.layer, service2 <- ..., } yield Service(service1, service2, repository1)`

Reply:
- Domain should rely on none of the packages, application can rely on domain, infrastructure can rely on domain and application and UI can depend on everything

Reply:
- Hexagonal architecture aims to make the application more flexible and easily testable
   - but it not main business concept that hexagone fix
      - suppose that u have microservice that is communicating with other ms with http
         - business concept is like that: 1st ms is ShoppingCartMicroservice and by API u buy product like GymEquipment and ShoppingCartMicroservice notifies ProductMicroservice that product was bought so we have 1 less in storage
            - load is very big so we decide to not use synchronous communication instead we use async like message broker like kafka
            - now the single thing u should change is change ProductRepositoryImplementation to support kafka instead of http; ProductRepository.get (when u get by http or gRPC) => ProductRepositoryGrpcImplementation, ProductRepositoryHttpImplementation

Reply:
- The business problem solved is that implementation details can change but the core business logic can remain untouched.

Reply:
- Pros : Testing is easier because you can easily create mock for example, changing implementation details like the underlying db is also much easier, and there is a clear separation of concern so the code base might be easier to understand
   - maybe not `underlying db` as it is never happening in practice, but changing protocol of communication like HTTP -> gRPC (much faster protocol) happens all the time, and changing way of communication like sync -> async (to decouple microservice) happens all the time

Reply:
- Cons : Might be an overkill for small project because of the numerous abstractions. Implementation can also be tricky from what I've read (not too sure why)
   - it partially true, when u make hexagon on every layer u get number of classes exploded in the project and it starts to pollute your codebase without any significant gain HOWEVER doing hexagon on very precise places will enable easy testing and switching protocols communications easily

Reply:
- It is solving the expression problem, which is adding new data variants or new operations without having to modify the existing code.
   - it is more about libraries not business services
   - tagless final is a way to 
      1. abstract over effect type: IO vs Future vs CompletableFuture: CustomerService[F[_]] and F is effect type
      2. switch interpreters of general business flows like `CustomerRepository.create(customer)(implicit customerRepositoryInterpreter) = customerRepositoryInterpreter.create(customer)`
         - in functional programming u don't have classes at all, u have only data (pure data like hashmaps) and u have behaviours that are functions, and behaviours are not connected with data at all
         - in OOP => behaviours of specific data is in one place => class; u can have just implementations because u have class fields, u put `customerRepositoryInterpreter` as a field in service class

Reply:
- Cons : Can lead to boiler plate code, and debugging might be more difficult
   - main problem is abstracting over effect type `F[_]` really makes sense?
      - how many times ur using something different that `IO` that carries errors `MonadThrow`?
         - nearly never, so ZIO is exactly tagless final that bind `IO` as default and not changeable effect type